Report policy revision implemented by the proxy in Endpoint model

joestringer opened the pull request:
This PR plumbs the Repository revision and endpoints into the `NetworkPolicyDiscoveryService`, where these are cached and associated with `NetworkPolicyCache` versions. When Envoy receives the relevant resources and acknowledges them, it looks up in this cache to determine which policy revision is being acked and calls back into the endpoint objects to update a new field, `proxyPolicyRevision`.

For more details, see #3144.

Fixes: #3122
Fixes: #3144

joestringer committed:
envoy: Apply L7 network policies per endpoint

Since there is endpoint-specific configuration which affects policy,
this patch changes network policies to be applied per endpoint (via IPs)
rather than per identity. Do this by defining an interface of the
relevant portions of the endpoint that are used when updating network
policy (NetworkPolicyEndpoint), and simplify UpdateNetworkPolicy() and
RemoveNetworkPolicy() to pass this new NetworkPolicyEndpoint around.

Fixes: #3122

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
endpoint: Add proxyPolicyRevision

This new field will report the policy revision that has been
acknowledged by the proxy, so it is clear to see whether the proxy is
synchronised with Cilium or not.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
api: Update golang comments

A while ago, the spec was updated to use double graves to indicate
fields, to be compatible with sphinx notation for where these fields end
up in the documentation. We didn't update the API definitions at the
time. Do so. This should be a functional no-op.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
envoy: Update revision upon acks from NPDS

When the NetworkPolicyDiscoveryService acknowledges oustanding
NetworkPolicyUpdates for an endpoint, push the corresponding policy
revision number into the endpoint.

Fixes: #3144

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
policy: Store policy revision in L4Policy

An upcoming commit will need to identify which policy revision was used
to generate an L4/L7 policy, so when we construct the L4Policy, cache
the current policy repository revision in the L4Policy object.

Signed-off-by: Joe Stringer <joe@covalent.io>

rlenglet commented on api/v1/openapi.yaml:
```diff
@@ -823,6 +823,9 @@ definitions:
         description: The policy revision this endpoint is running on
         type: integer
+      proxy-policy-revision:
+        description: The policy revision currently in use in the proxy
```
in use -> enforced (?)

rlenglet commented on pkg/envoy/endpoint.go:
```diff
@@ -0,0 +1,167 @@
+)
+
+var (
+	defaultipToBindingSize = 64
```
const

rlenglet replied:
ip -> IP

rlenglet commented on pkg/envoy/endpoint.go:
```diff
@@ -0,0 +1,167 @@
+	GetIdentity() identity.NumericIdentity
+
+	// OnProxyPolicyUpdate is called when the proxy confirms that it
+	// has implemented a policy.
```
implemented -> applied

rlenglet commented on pkg/envoy/endpoint.go:
```diff
@@ -0,0 +1,167 @@
+
+var (
+	defaultipToBindingSize = 64
+	unackedNPDSUpdates     *NetworkPolicyEventList
```
Add a TODO(rlenglet) comment: move this into the XDSServer.

rlenglet commented on pkg/envoy/endpoint.go:
```diff
@@ -0,0 +1,167 @@
+	}
+}
+
+// Remove an endpoint from the list so that it no longer receives callbacks
```
Remove removes ...

rlenglet commented on pkg/envoy/endpoint.go:
```diff
@@ -0,0 +1,167 @@
+// revision and a NetworkPolicyCache version for a particular endpoint.
+type NetworkPolicyEventList struct {
+	lock.Mutex
+	ipToBinding map[string][]versionBinding
```
IPv6 addresses may have different representations. Make sure to normalize the address representations before using them as keys.

rlenglet commented on pkg/envoy/endpoint.go:
```diff
@@ -0,0 +1,167 @@
+	npel.Lock()
+	defer npel.Unlock()
+
+	revisions := npel.ipToBinding[addr]
```
Check whether that key has been found.
Log at Warn level, and return.

rlenglet commented on pkg/envoy/endpoint.go:
```diff
@@ -0,0 +1,167 @@
+	for _, r := range revisions {
+		if r.cacheVersion > version {
+			newRevisions = append(newRevisions, r)
+		} else if r.cacheVersion == version {
```
Remove the condition:
```
} else {
```

rlenglet commented on pkg/envoy/endpoint.go:
```diff
@@ -0,0 +1,167 @@
+
+// getRevisionBinding fetches the policy revision and endpoint pointer
+// corresponding to the specified address and NPDS cache version.
+func (npel *NetworkPolicyEventList) getRevisionBinding(addr string, version uint64) (policyRevision uint64, ep NetworkPolicyEndpoint) {
```
getRevisionBinding -> getRevisionBindings
Return a slice of all the policy revisions that have a cache version <= version, along with the (optional) Completion. Maybe just return a `[]revisionBinding`.

rlenglet commented on pkg/envoy/endpoint.go:
```diff
@@ -0,0 +1,167 @@
+	cacheVersion   uint64
+	policyRevision uint64
+	endpoint       NetworkPolicyEndpoint
+}
```
Add an optional `completion.Completion` in there.

rlenglet commented on pkg/envoy/endpoint.go:
```diff
@@ -0,0 +1,167 @@
+
+	addr, err := xds.IstioNodeToIP(node)
+	if err != nil {
+		log.Debugf("Recieved invalid address from Istio: %v", node)
```
Recieved -> received
Istio -> Envoy proxy

rlenglet commented on pkg/envoy/endpoint.go:
```diff
@@ -0,0 +1,167 @@
+		return
+	}
+
+	policyRevision, ep := unackedNPDSUpdates.getRevisionBinding(addr, version)
```
We need to have a different behavior depending on whether we run Envoy on the host or in sidecar proxies, based on the `sidecar-http-proxy` flag (https://github.com/cilium/cilium/pull/3126).
In case that flag is false, the node IP address will always be `127.0.0.1`, because all endpoints are handled by the local Envoy proxy. In that case, you need to "wildcard" the node address in `getRevisionBindings`, i.e. only match bindings based on cache version.

rlenglet replied:
Call all the returned non-nil `Completion`s, in addition to calling `OnProxyPolicyUpdate` below.

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -506,22 +506,46 @@ func getNetworkPolicy(id identity.NumericIdentity, policy *policy.L4Policy,
-func UpdateNetworkPolicy(id identity.NumericIdentity, policy *policy.L4Policy,
+// When the proxy acknowledges the network policy update, it will result in
+// a subsequent call to the endpoint's OnProxyPolicyAcknowledge() function.
+func UpdateNetworkPolicy(ep NetworkPolicyEndpoint, policy *policy.L4Policy,
```
Add a `wg *completion.WaitGroup` parameter.

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -506,22 +506,46 @@ func getNetworkPolicy(id identity.NumericIdentity, policy *policy.L4Policy,
-	name := strconv.FormatUint(uint64(id), 10)
-	NetworkPolicyCache.Upsert(NetworkPolicyTypeURL, name, networkPolicy, false)
+	// When successful, push them into the cache.
+	unackedNPDSUpdates.Insert(ep, policy.Revision, NetworkPolicyCache.GetVersion())
```
Call `wg.AddCompletion()` to get a `Completion` and pass that to the `Insert` call to store it in the binding.

rlenglet commented on daemon/daemon.go:
```diff
@@ -148,21 +148,21 @@ func (d *Daemon) RemoveProxyRedirect(e *endpoint.Endpoint, id string) error {
 // UpdateNetworkPolicy adds or updates a network policy in the set
 // published to L7 proxies.
-func (d *Daemon) UpdateNetworkPolicy(id identity.NumericIdentity, policy *policy.L4Policy,
+func (d *Daemon) UpdateNetworkPolicy(e *endpoint.Endpoint, policy *policy.L4Policy,
```
Add a `wg *completion.WaitGroup` parameter.

joestringer replied:
Done.

rlenglet commented on envoy/cilium/npds.proto:
```diff
@@ -47,6 +47,11 @@ message NetworkPolicy {
+  // The policy identifier associated with the network policy. Corresponds to
+  // remote_policies entries in PortNetworkPolicyRule.
+  // Optional.
+  uint64 policy = 4;
```
Just a nitpick. I would move this up and use the following labels:
```
string name = 1;
uint64 policy = 2;
repeated PortNetworkPolicy ingress_per_port_policies = 3;
repeated PortNetworkPolicy egress_per_port_policies = 4;
```

rlenglet commented on pkg/endpoint/endpoint.go:
```diff
@@ -1883,6 +1885,17 @@ func (e *Endpoint) bumpPolicyRevision(revision uint64) {
 }
 
+// OnProxyPolicyUpdate is a callback used to update the Endpoint's
+// proxyPolicyRevision when the specified revision has been implemented in the
```
implemented -> applied

rlenglet commented on pkg/endpoint/owner.go:
```diff
@@ -53,12 +53,12 @@ type Owner interface {
 	// UpdateNetworkPolicy adds or updates a network policy in the set
 	// published to L7 proxies.
-	UpdateNetworkPolicy(id identity.NumericIdentity, policy *policy.L4Policy,
+	UpdateNetworkPolicy(e *Endpoint, policy *policy.L4Policy,
```
Add a `wg *completion.WaitGroup` parameter.

joestringer replied:
Done.

joestringer committed:
completion: Add a callback upon completion

This callback will be called when the completion socket is closed. It
allows an interested party to immediately apply some behaviour at the
time that the completion occurs, rather than waiting on the completion
in the outer code. It will be used in an upcoming commit to reflect the
policy revision from proxy acknowledgements up to the endpoint.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
endpoint: Fix typo

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
endpoint: Update L7 Network Policy later

The existing path was updating the L7 Network Policies in the
potentially dry-run path inside regeneratePolicy(), however it should
only be applied when we are actually going to apply the policy changes
to the datapath. Refactor the function and call it later from
regenerateBPF().

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer commented on daemon/daemon.go:
```diff
@@ -148,21 +148,22 @@ func (d *Daemon) RemoveProxyRedirect(e *endpoint.Endpoint, id string) error {
 		return fmt.Errorf("can't update network policy, proxy disabled")
 	}
-	return d.l7Proxy.UpdateNetworkPolicy(id, policy, labelsMap, deniedIngressIdentities, deniedEgressIdentities)
+	// TODO(jrajahalme): Pass e.ProxyWaitGroup down
```
+ @jrajahalme you'll need to pass `e.ProxyWaitGroup` down as part of your changes.

rlenglet replied:
@jrajahalme heads up. We can't pass a WaitGroup yet, because no proxy is querying + acking the NetworkPolicy resources yet, so all regenerations would fail if we did.

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -506,22 +507,63 @@ func getNetworkPolicy(id identity.NumericIdentity, policy *policy.L4Policy,
+		ep.GetIPv4Address(),
+	}
+	policies := []*cilium.NetworkPolicy{}
+	for i := range ips {
```
Nitpick: `for _, ip := range ips {` then use `ip` instead of `ips[i]`.

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -506,22 +507,63 @@ func getNetworkPolicy(id identity.NumericIdentity, policy *policy.L4Policy,
-	name := strconv.FormatUint(uint64(id), 10)
-	NetworkPolicyCache.Upsert(NetworkPolicyTypeURL, name, networkPolicy, false)
+	// When successful, push them into the cache.
+	for i := range policies {
```
Nitpick: `for _, p := range policies {` then use `p` instead of `policies[i]`.

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -506,22 +507,63 @@ func getNetworkPolicy(id identity.NumericIdentity, policy *policy.L4Policy,
+		} else {
+			c = wg.AddCompletionWithCallback(callback)
+		}
+		nodeIDs := []string{}
```
Nitpick:
```
nodeIDs := make([]string, 0, 1)
```

rlenglet commented on pkg/envoy/xds/cache.go:
```diff
@@ -254,3 +254,8 @@ func (c *Cache) GetResources(ctx context.Context, typeURL string, lastVersion *u
 }
+
+// GetVersion returns the version for the cache at the current point in time.
+func (c *Cache) GetVersion() uint64 {
```
Remove. That seems unused.
Anyway it is not thread-safe as-is. This would require locking.

rlenglet reviewed (approved):
LGTM. Thanks for doing this! I just have 4 nitpicks. Nothing blocking.
